pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -301,10 +328,48 @@ func (c *Controller) onServiceEvent(curr interface{}, event model.Event) error {
 	return nil
 }
 
-func (c *Controller) onNodeEvent(obj interface{}, ev model.Event) error {
+// TODO: move to API
+// The value for this annotation is a set of key value pairs (node labels)
+// that can be used to select a subset of nodes from the pool of k8s nodes
+const nodeSelectorAnnotation = "traffic.istio.io/nodeSelector"
+
+func getNodeSelectorsForService(svc v1.Service) labels.Instance {
+	if svc.Annotations != nil && svc.Annotations[nodeSelectorAnnotation] != "" {
+		var nodeSelectorKV map[string]string
+		if err := json.Unmarshal([]byte(svc.Annotations[nodeSelectorAnnotation]), &nodeSelectorKV); err != nil {
+			log.Debugf("failed to unmarshal node selector annotation value for service %s.%s: %v",
+				svc.Name, svc.Namespace, err)
+		}
+		return nodeSelectorKV
+	}
+	return nil
+}
+
+func (c *Controller) onNodeMetadataEvent(obj interface{}, event model.Event) error {
 	return c.checkReadyForEvents()
 }
 
+func (c *Controller) onNodeEvent(obj interface{}, event model.Event) error {
+	if err := c.checkReadyForEvents(); err != nil {
+		return err
+	}
+
+	// TODO: optimize me. We just need a full EDS push
```

Comment:
This caused huge regression: https://github.com/istio/istio/issues/22963. Nodes change all the time, like every 5s (on GKE at least). At the very very least we need to do some logic here to filter out events we care about like we do with all other types. Its also retriggered for all nodes every 60s due to resync period. In 1 min with just 10 nodes this triggered 60 full pushes (after debounce)